File: skills/docforge/_shared/runtime/common/js/evidence_locators.js
"use strict";
/** Validate immutable implementation evidence locators in generated Markdown. */

const fs = require("fs");
const path = require("path");
const pf = require("./provenance_frontmatter.js");
const { lineCount, rangeBlobHash, rawBlobHash } = require("./evidence_hash.js");

const LOCATOR = /([A-Za-z0-9][A-Za-z0-9_./-]*)#L([1-9]\d*)-L([1-9]\d*) @ ([0-9a-f]{40})/g;
const HEADING = /^(#{1,6})\s+(.*\S)\s*$/;
function anchor(value) { return value.toLowerCase().replace(/[^\p{L}\p{N}_\s-]/gu, "").trim().replace(/[\s-]+/g, "-"); }
function root(filePath) {
  let current = path.dirname(filePath);
  while (true) {
    if (fs.existsSync(path.join(current, ".git")) || fs.existsSync(path.join(current, ".docforge"))) return current;
    const parent = path.dirname(current);
    if (parent === current) return path.dirname(filePath);
    current = parent;
  }
}
function outsideFences(text) {
  const lines = []; let inFence = false; let marker = "";
  for (const [index, line] of text.split(/\r?\n/).entries()) {
    const match = line.match(/^\s*(`{3,})/);
    if (match) { if (!inFence) { inFence = true; marker = match[1]; } else if (match[1] === marker) { inFence = false; marker = ""; } continue; }
    if (!inFence) lines.push([index + 1, line]);
  }
  return lines;
}
/* `provenance` and `bodyStart` come from the caller's already-resolved
 * metadata. Passing them is what makes this work under sidecar storage, where
 * the markdown carries no frontmatter to parse — without them the check would
 * find no provenance and silently pass every document. When they are omitted
 * the frontmatter is parsed directly, for callers holding a legacy document. */
function validateLocators(document, text = null, provenance = null, bodyStart = null) {
  document = path.resolve(document);
  const contents = text == null ? fs.readFileSync(document, "utf8") : text;
  if (provenance === null) {
    const parsed = pf.parseFrontmatter(contents);
    if (parsed.state !== "ok" || !parsed.provenance || typeof parsed.provenance !== "object") return [];
    provenance = parsed.provenance;
    if (bodyStart === null) bodyStart = parsed.end;
  }
  if (!provenance || typeof provenance !== "object") return [];
  if (bodyStart === null) bodyStart = 0;
  const headings = [];
  // Newlines before the body, so the first body line is 1-based like the
  // locator scan below. `split().length` would count lines, not newlines.
  const bodyLine0 = contents.slice(0, bodyStart).split(/\r?\n/).length - 1;
  for (const [index, line] of contents.slice(bodyStart).split(/\r?\n/).entries()) {
    const match = line.match(HEADING); if (match) headings.push([bodyLine0 + index + 1, anchor(match[2])]);
  }
  const sourcePairs = new Map();
  for (const section of provenance.sections || []) {
    sourcePairs.set(section.id, new Set((section.sources || []).map((source) => `${source.path}\0${source.git_blob}`)));
  }
  const repo = root(document); const defects = [];
  for (const [lineNumber, line] of outsideFences(contents)) {
    LOCATOR.lastIndex = 0; let match;
    while ((match = LOCATOR.exec(line)) !== null) {
      const [, rel, startRaw, endRaw, digest] = match; const start = Number(startRaw); const end = Number(endRaw);
      if (path.posix.isAbsolute(rel) || rel.split("/").includes("..")) { defects.push({ kind: "evidence path escape", line: lineNumber, detail: rel }); continue; }
      const target = path.resolve(repo, ...rel.split("/"));
      if (!fs.existsSync(target) || !fs.statSync(target).isFile() || !(target === repo || target.startsWith(`${repo}${path.sep}`))) { defects.push({ kind: "evidence source missing", line: lineNumber, detail: rel }); continue; }
      const bytes = fs.readFileSync(target);
      const wholeFile = rawBlobHash(bytes);
      const scoped = rangeBlobHash(bytes, start, end);
      if (digest !== wholeFile && digest !== scoped) defects.push({ kind: "stale evidence blob", line: lineNumber, detail: rel });
      const count = lineCount(bytes);
      if (end < start || count === null || end > count) defects.push({ kind: "invalid evidence range", line: lineNumber, detail: `${rel}#L${start}-L${end}` });
      const heading = [...headings].reverse().find(([headingLine]) => headingLine <= lineNumber);
      if (!heading) defects.push({ kind: "unknown evidence heading", line: lineNumber, detail: rel });
      else if (!(sourcePairs.get(heading[1]) || new Set()).has(`${rel}\0${digest}`)) defects.push({ kind: "evidence provenance mismatch", line: lineNumber, detail: rel });
    }
  }
  return defects;
}
module.exports = { validateLocators };

File: skills/docforge/_shared/runtime/common/python/evidence_locators.py
"""Validate immutable implementation evidence locators in generated Markdown."""

from __future__ import annotations

import re
from pathlib import Path

from runtime.common.python.provenance_frontmatter import parse_frontmatter
from runtime.common.python.evidence_hash import line_count, range_blob_hash, raw_blob_hash

LOCATOR_RE = re.compile(r"(?P<path>[A-Za-z0-9][A-Za-z0-9_./-]*)#L(?P<start>[1-9]\d*)-L(?P<end>[1-9]\d*) @ (?P<blob>[0-9a-f]{40})")
HEADING_RE = re.compile(r"^(#{1,6})\s+(.*\S)\s*$")


def _anchor(value: str) -> str:
    return re.sub(r"[\s-]+", "-", re.sub(r"[^\w\s-]", "", value.lower()).strip()).strip("-")


def _root(path: Path) -> Path:
    for parent in [path.parent, *path.parents]:
        if (parent / ".git").exists() or (parent / ".docforge").exists():
            return parent
    return path.parent


def _outside_fences(text: str) -> list[tuple[int, str]]:
    lines: list[tuple[int, str]] = []
    fence = False
    marker = ""
    for number, line in enumerate(text.splitlines(), 1):
        match = re.match(r"^\s*(`{3,})", line)
        if match:
            if not fence:
                fence, marker = True, match.group(1)
            elif match.group(1) == marker:
                fence, marker = False, ""
            continue
        if not fence:
            lines.append((number, line))
    return lines


def validate_locators(document: Path, text: str | None = None) -> list[dict]:
    """Return structured locator defects without broadening provenance 2.0."""
    text = document.read_text(encoding="utf-8", errors="replace") if text is None else text
    document = document.resolve()
    state, provenance, body_start = parse_frontmatter(text)
    if state != "ok" or not isinstance(provenance, dict):
        return []
    body = text[body_start:]
    body_line0 = text[:body_start].count("\n")
    headings: list[tuple[int, str]] = [
        (body_line0 + index, _anchor(match.group(2)))
        for index, line in enumerate(body.splitlines(), 1)
        if (match := HEADING_RE.match(line))
    ]
    source_pairs = {
        section.get("id"): {
            (source.get("path"), source.get("git_blob"))
            for source in section.get("sources", []) if isinstance(source, dict)
        }
        for section in provenance.get("sections", []) if isinstance(section, dict)
    }
    root = _root(document)
    defects: list[dict] = []
    for line_number, line in _outside_fences(text):
        for match in LOCATOR_RE.finditer(line):
            rel, start, end, digest = match.group("path"), int(match.group("start")), int(match.group("end")), match.group("blob")
            if ".." in Path(rel).parts or Path(rel).is_absolute():
                defects.append({"kind": "evidence path escape", "line": line_number, "detail": rel})
                continue
            target = (root / rel).resolve()
            if not target.is_file() or root not in [target, *target.parents]:
                defects.append({"kind": "evidence source missing", "line": line_number, "detail": rel})
                continue
            actual = target.read_bytes()
            whole_file = raw_blob_hash(actual)
            scoped = range_blob_hash(actual, start, end)
            if digest != whole_file and digest != scoped:
                defects.append({"kind": "stale evidence blob", "line": line_number, "detail": rel})
            count = line_count(actual)
            if end < start or count is None or end > count:
                defects.append({"kind": "invalid evidence range", "line": line_number, "detail": f"{rel}#L{start}-L{end}"})
            heading = next((anchor for heading_line, anchor in reversed(headings) if heading_line <= line_number), None)
            if heading is None:
                defects.append({"kind": "unknown evidence heading", "line": line_number, "detail": rel})
            elif (rel, digest) not in source_pairs.get(heading, set()):
                defects.append({"kind": "evidence provenance mismatch", "line": line_number, "detail": rel})
    return defects
